middleware/kubernetes/ns.go

```diff
@@ -0,0 +1,111 @@
+package kubernetes
+
+import (
+	"net"
+	"strings"
+
+	"github.com/coredns/coredns/middleware/etcd/msg"
+
+	"github.com/miekg/dns"
+	"k8s.io/client-go/1.5/pkg/api"
+)
+
+const DefaultNSName = "ns.dns."
+
+var corednsRecord dns.A
+
+func (k *Kubernetes) recordsForNS(r recordRequest, svcs *[]msg.Service) error {
+	ns := k.CoreDNSRecord()
+	s := msg.Service{
+		Host: ns.A.String(),
+		Key:  msg.Path(strings.Join([]string{ns.Hdr.Name, r.zone}, "."), "coredns")}
+	*svcs = append(*svcs, s)
+	return nil
+}
+
+// DefaultNSMsg returns an msg.Service representing an A record for
+// ns.dns.[zone] -> dns service ip. This A record is needed to legitimize
+// the SOA response in middleware.NS(), which is hardcoded at ns.dns.[zone].
+func (k *Kubernetes) DefaultNSMsg(r recordRequest) msg.Service {
+	ns := k.CoreDNSRecord()
+	s := msg.Service{
+		Key:  msg.Path(strings.Join([]string{DefaultNSName, r.zone}, "."), "coredns"),
+		Host: ns.A.String(),
+	}
+	return s
+}
+
+func IsDefaultNS(name string, r recordRequest) bool {
+	return strings.Index(name, DefaultNSName) == 0 && strings.Index(name, r.zone) == len(DefaultNSName)
+}
+
+type ifAddrsFunc func() ([]net.Addr, error)
+
+func (k *Kubernetes) CoreDNSRecord() dns.A {
+	return k.DoCoreDNSRecord(net.InterfaceAddrs)
+}
+
+func (k *Kubernetes) DoCoreDNSRecord(ifAddrs ifAddrsFunc) dns.A {
```

Comment:
That was for quick and dirty stubbing.  I have replaced it with a more idiomatic struct wrapper and interface.